participant/app/components/FileUploader.tsx

```diff
@@ -0,0 +1,185 @@
+import { Trans, useTranslation } from "react-i18next";
+import React, {
+  useState,
+  useRef,
+  useImperativeHandle,
+  useEffect,
+  forwardRef,
+} from "react";
+import type { i18nKey } from "~/types";
+import type { FileInputRef } from "app/components/internal/FileInput";
+export type { FileInputRef } from "app/components/internal/FileInput";
+
+export type FileUploaderProps = {
+  className?: string;
+  id: string;
+  name: string;
+  labelKey: i18nKey;
+  accept?: string;
+  required?: boolean;
+  maxFileSizeInBytes?: number;
+  maxFileCount?: number;
+};
+
+export type FileState = {
+  [key: string]: File;
+};
+
+export type FileError = {
+  errorType?: "count" | "size" | "type";
+};
+
+export const FileUploaderForwardRef: React.ForwardRefRenderFunction<
+  FileInputRef,
+  FileUploaderProps & JSX.IntrinsicElements["input"]
+> = (
+  {
+    className,
+    id,
+    name,
+    labelKey,
+    accept,
+    required,
+    maxFileSizeInBytes = 26_214_400,
+    maxFileCount = 20,
+    ...inputProps
+  },
+  ref
+): React.ReactElement => {
+  const { t } = useTranslation();
+  const [files, setFiles] = useState<FileState>({});
+  const [previews, setPreviews] = useState(<></>);
+  const fileInputRef = useRef(null);
+  const documentString = t(`${labelKey}.document`);
+  const [errorMessage, setErrorMessage] = useState("");
+
+  const ifNeeded = t(`${labelKey}.ifNeeded`);
+  const currentDocumentNumber = Object.keys(files).length;
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      input: fileInputRef.current,
+      files: Object.keys(files).map((key) => files[key]),
+    }),
+    [files]
+  );
+
+  const renderPreviews = () => {
+    setPreviews(
+      <>
+        {Object.keys(files).map((fileName, index) => {
+          let file = files[fileName];
+          return (
+            <div key={`document-${index + 1}`}>
+              <span>{`${documentString} ${index + 1}`}</span>
+              <FilePreview
+                imageId={`preview-${index}`}
+                file={file}
+                clickHandler={removeFile}
+                removeFileKey={`${labelKey}.removeFile`}
+                selectedKey={`${labelKey}.selected`}
+                altTextKey={`${labelKey}.altText`}
+              />
+            </div>
+          );
+        })}
+      </>
+    );
+  };
+  // eslint-disable-next-line   react-hooks/exhaustive-deps -- (deps list is correct)
+  useEffect(renderPreviews, [files, documentString, labelKey]);
+  const fileTypeCheck = (newFile: File): boolean => {
+    if (!accept || accept == "*") {
+      return true;
+    }
+    const acceptedTypes = accept.split(",");
+    let fileTypeAllowed = true;
+    if (fileTypeAllowed) {
+      for (let j = 0; j < acceptedTypes.length; j += 1) {
+        const fileType = acceptedTypes[parseInt(`${j}`)];
+        fileTypeAllowed =
+          newFile.name.indexOf(fileType) > 0 ||
+          newFile.type.includes(fileType.replace(/\*/g, ""));
```

Comment:
Can you walk me through what this does? What is the expected structure of `newFile`?

Reply:
The structure? Or the type? It's a `File` object